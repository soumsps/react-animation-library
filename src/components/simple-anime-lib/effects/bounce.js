import React, { useRef, memo } from 'react';
import { DEFAULT_BOUNCE_CONFIG } from '../constants';
import {
  useAnimationFrame,
  timingFunction,
  getDelay,
  getDuration,
  getIterationCount,
  getEasing,
  getDropHeight,
  getCallback,
} from '../helper';

const Bounce = memo((props) => {
  const delay = getDelay(props.delay, DEFAULT_BOUNCE_CONFIG.delay);
  const duration = getDuration(props.duration, DEFAULT_BOUNCE_CONFIG.duration);
  const iterationCount = getIterationCount(
    props.iterationCount,
    DEFAULT_BOUNCE_CONFIG.iterationCount
  );
  const easing = getEasing(props.easing, DEFAULT_BOUNCE_CONFIG.easing);
  const callback = getCallback(props.callback, DEFAULT_BOUNCE_CONFIG.callback);
  const dropHeight = getDropHeight(
    props.dropHeight,
    DEFAULT_BOUNCE_CONFIG.dropHeight
  );
  const timing = timingFunction[easing];

  const elementRef = useRef({
    style: { position: 'relative', top: `0px` },
  });

  const draw = (progress) => {
    if (elementRef.current !== null) {
      if (progress < 0.2) {
        elementRef.current.style.top = dropHeight * (1 / 0.2) * progress + 'px';
      } else {
        elementRef.current.style.top = dropHeight * (1 - progress) + 'px';
      }
    }
  };

  useAnimationFrame({
    delay,
    duration,
    callback,
    iterationCount,
    timing,
    draw,
  });

  return (
    <span ref={elementRef} style={elementRef.current.style}>
      {props.children}
    </span>
  );
});

export { Bounce };
