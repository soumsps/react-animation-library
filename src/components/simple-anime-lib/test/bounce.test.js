import React from 'react';
import renderer from 'react-test-renderer';
import { Bounce } from '../bounce';

it('bounce effect snapshot test', () => {
  const tree = renderer.create(<Bounce></Bounce>).toJSON();
  expect(tree).toMatchSnapshot();
});
