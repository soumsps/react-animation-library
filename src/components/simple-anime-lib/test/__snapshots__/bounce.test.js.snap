// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bounce effect snapshot test 1`] = `
<span
  style={
    Object {
      "position": "relative",
      "top": "0px",
    }
  }
/>
`;
